Add support for Gradle 9

vierbergenlars opened the pull request:
Conventions are removed from Gradle 9 and some previously-deprecated methods have been removed as well.

To be able to write the code supporting the replacement methods, I had to update the build to Gradle 8.3 as well.
This required updating the plugin-publish plugin and updating the build script a bit.

To be able to run an integration test against Gradle 9, I have to run the test JVM with Java 17 (because Java 17 is required to run Gradle 9 itself), but to run tests for older Gradle versions, I needed a JVM with Java 11 (because those versions don't support running with Java 17 yet).

To make it easier to run the tests against multiple Java versions locally, I split up the integration test tasks in two, each running with their own Java toolchain.

Fixes https://github.com/reyerizo/jcstress-gradle-plugin/issues/50

vierbergenlars committed:
Upgrade to Gradle 8

vierbergenlars committed:
Fix plugin to allow running with Gradle 9

vierbergenlars committed:
Let Gradle plugin generate plugin metadata file

vierbergenlars committed:
Run tests with Java 8 to support older Gradle versions

vierbergenlars committed:
Run integration tests with Java 11 and Java 17 using toolchains

The integration test for Gradle 5 requires Java 11, and the test for Gradle 9 requires Java 17

JarvisCraft commented:
Hi there, @jerzykrlk, could you please take a look at this PR?
This is currently a blocker for codebases moving to Gradle 9+ (which is also a requirement for Java 25 builds) so a release with this fix would be of significant importance!

jerzykrlk commented:
Hi Petr, Lars! Thanks for the contribution! 

I've launched the CI and saw some tests failing - would you be able to check them?

I'll try to debug them locally, but I'm not sure if I can make it this week. Apologies.

vierbergenlars committed:
Update spec tests for Gradle 8

There were some removed methods that were still being used in the spec tests which are no longer present in Gradle 8.

vierbergenlars commented:
I somehow missed the fact that there also was a non-integration tests.
Because they are now running with Gradle 8, some previously-existing properties no longer exist, and I have updated those as well.

eortega-pjr commented:
I just ran into this and am happy to see a resolution is already in progress. Thanks.

JarvisCraft commented:
All credits go to @vierbergenlars, I was just here to ping you :)
Thanks!

vierbergenlars committed:
Add dependency between compileIntegrationGroovy and createPluginClasspath

jerzykrlk commented:
Hey @vierbergenlars — I just published version 0.9.0 to the Gradle Plugin Portal. It should be good to go now — could you check if it works for you?

Thanks again for the PR!

By the way, I’ll probably do a small cleanup and release another version in a few weeks. I’ve run into some odd build issues on Windows 11 and noticed a tiny number formatting regression.
